phantom-drift/introspection.proto

```diff
@@ -0,0 +1,351 @@
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
+
+package introspection.pb;
+
+// ResultCounter is a monotonically increasing counter that reports an ok/err breakdown of the total.
+message ResultCounter {
+    uint32 total = 1;
+    uint32 ok = 2;
+    uint32 err = 3;
+}
+
+// Moving totals over different time windows. Not all have to be filled.
+message TimeWindowCounter {
+    uint32 over_1m      = 1;
+    uint32 over_5m      = 2;
```

Comment:
I am not sure about this model. For example, a connection of 16 minutes should be accounted for `over_15m`, `over_30m`, ...? 

I was thinking that this would be a more flexible type. Like a list of times for every connection / stream. Why should we go in this way?

Reply:
> a connection of 16 minutes should be accounted for over_15m, over_30m, ...?

The "connection opened" event for a connection that was opened 16 minutes ago should appear in the over_30m and above (over_1h, over_2h...) buckets. I think what's confusing is the term "over" which here it stands for "window over 1h", not "quantity over".

What I'm thinking is that, when you connect, you want a sense of what the system has been up to in the recent past, despite not having any details of that activity (e.g. past streams, etc.) (for that you'd need a recording).

I don't think instrumentation libs like OC or Prometheus give us this functionality OOTB. Those range totals are calculated on the database. But we can implement this striped counter with a circular array; assuming a super naïve solution with a resolution of 1min, each metric would require 1440 buckets of uint32, we're talking ~6kb per metric, with efficient access. We can get more sophisticated though.

Reply:
To illustrate this idea, here's some ascii art where X represents the event at T-16min:

```
time    past -> present
======================================================X================
                                                      |               | 1m
                                                      |           |---| 5m  
                                                      | |-------------| 15m
                                         |------------X---------------| 30m
           |------------------------------------------X---------------| 60m
```

Reply:
> What I'm thinking is that, when you connect, you want a sense of what the system has been up to in the recent past, despite not having any details of that activity (e.g. past streams, etc.) (for that you'd need a recording).

fair enough, I agree!